### DISCUSSION TITLE
matchAll returning all matching groups

### DISCUSSION THREAD

**Discussion by hebinet** [laravel/framework#42529] [Ideas]
In my Code I use something like this:
```php
Str::of($job->location->address)
            ->trim()
            ->matchAll('/^(.*?),\s+(\d+)\s+(.*)$/')
```
and only getting the first group match back.

Wouldn't it be better that all the matching groups should be returned?

This would be easily accomplished by changing the `matchAll()` method return in Str class from 
```php
if (empty($matches[0])) {
    return collect();
}

return collect($matches[1] ?? $matches[0]);
```

to 

```php
if (empty($matches[0])) {
    return collect();
}

if (count($matches === 1)) {
    return collect($matches[0]);
}

return collect(array_slice($matches, 1));
```

What do you think?
This would be propably a breaking change and therefore could only be implemented in v10+.

**Comment by django23** [comment]
Agree, or at least a documentation update that explains it return only the first group.

**Comment by shaedrich** [comment]
Aside from that, `return collect($matches[1] ?? $matches[0]);` looks [a little counterintiuitve](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) to me

**Reply by shaedrich** [reply]
Well, I have a hunch as to why this was done, but having this as the first entry of an array is a limitation of the language itself, yet, it doesn't have to be one of the framework. The method could return a class, that would return the matches, starting at 1, via `__get()` and has `$matches[0]` as a named property.

But yeah, that would be even more of a breaking change …
